storages/python/python-storage-plugins/bson.py

```diff
@@ -0,0 +1,170 @@
+"""A DLite storage plugin for BSON written in Python."""
+import os
+import subprocess
+import sys
+
+import binascii
+from binascii import unhexlify
+import bson as pybson # Must be pymongo.bson
+
+import dlite
+from dlite.options import Options
+from dlite.utils import instance_from_dict
+
+
+class bson(DLiteStorageBase):
+    """DLite storage plugin for BSON."""
+
+    def open(self, uri, options=None):
+        """Open `uri`.
+
+        The `options` argument provies additional input to the driver.
+        Which options that are supported varies between the plugins.
+        It should be a valid URL query string of the form:
+
+            key1=value1;key2=value2...
+
+        An ampersand (&) may be used instead of the semicolon (;).
+
+        Typical options supported by most drivers include:
+        - mode : append (default) | r | w
+            Valid values are:
+            - append   Append to existing file or create new file
+            - r        Open existing file for read-only
+            - w        Truncate existing file or create new file
+
+        After the options are passed, this method may set attribute
+        `writable` to True if it is writable and to False otherwise.
+        If `writable` is not set, it is assumed to be True.
+        
+        The BSON data is translated to JSON.
+        """
+        self.options = Options(options, defaults='mode=append')
+        self.mode = dict(r='rb', w='wb', append='rb+')[self.options.mode]
+        self.writable = False if 'rb' in self.mode else True
+        self.uri = uri
+        self.d = {}
+        if self.mode in ('rb', 'rb+') and os.path.exists(uri):
+            with open(uri, self.mode) as f:
+                bson_data = f.read()
+                if pybson.is_valid(bson_data):
+                    self.d = pybson.decode(bson_data)
+                    if self.d == {}:
+                        raise EOFError("Failed to read BSON data from " + uri)
+                else:
+                    raise EOFError("Invalid BSON data in source " + uri)
+
+    def close(self):
+        """Close this storage and write the data to file.
+
+        Assumes the data to store is in JSON format.
+        """
+        if self.writable:
```

Comment:
I am probably an ignorant, but I don't understand why we need these conversions.

I testet the following which works for me (prints True):
```python
import bson

import dlite
from dlite.utils import instance_from_dict


inst = dlite.get_instance(dlite.ENTITY_SCHEMA)

data = bson.encode(inst.asdict())
with open('raw.bson', 'wb') as f:
    f.write(data)

with open('raw.bson', 'rb') as f:
    newdata = f.read()
d = bson.decode(newdata)
inst2 = instance_from_dict(d)

print(inst == inst2)
```

Reply:
Is this metadata? The problem I encountered was with actual data (the 'myblob' properties in the file 'src/tests/test-data.json'). The DLite entity stores e.g. the 'myblob' value '0a0010' (for uuid 'e076a856-e36e-5335-967e-2f2fd153c17d') as "bytearray(b'\n\x00\x10')", and bson.encode() didn't accept that input, so it had to be translated. But translating it directly back to '0a0010' didn't work**, as DLite somehow caused the backslashes to be doubled, so that the translated hexadecimal value became '5c6e5c7830305c783130' (corresponding to b'\\\n\\\x00\\\x10').

** It worked when run through the Python interpreter, but not when run through DLite. This is why I wrote "The plugin is therefore written as a workaround. If this issue can be resolved, the plugin Python code can probably be simplified."